### PR TITLE
rbd-mirror: fix journal shut down ordering

### PR DESCRIPTION
**Pull request by dillaman**

### CODE REVIEW

**Comment by dillaman** [comment]
retest this please

File: src/librbd/Journal.cc [hunk 1]
```diff
@@ -1069,19 +1069,51 @@ typename Journal<I>::Future Journal<I>::wait_event(Mutex &lock, uint64_t tid,
...
+  assert(m_state == STATE_READY);
+  assert(m_journal_replay == nullptr);
+
+  if (r < 0) {
+    lderr(cct) << "failed to stop recording: " << cpp_strerror(r) << dendl;
+    *journal_replay = nullptr;
+
+    // get back to a sane-state
+    start_append();
+    on_finish->complete(r);
```
Comment (trociny) [inline src/librbd/Journal.cc]: @dillaman Is it safe to complete under the lock? (here and below)

Comment (dillaman) [inline src/librbd/Journal.cc]: Yes, in ```start_external_replay``` I wrapped the provided ```on_finish``` callback with an async wrapper that will pop it on the op_work_queue when completed.

**Comment by trociny** [comment]
lgtm, teuthology: http://pulpito.ceph.com/trociny-2016-06-09_09:52:22-rbd-wip-mgolub-testing---basic-vps/

**Pull request merged**